src/Umbraco.Web/Routing/UrlProvider.cs

```diff
@@ -210,17 +208,20 @@ public string GetUrl(int id, UrlProviderMode mode)
         /// </remarks>
         public string GetUrl(int id, Uri current, UrlProviderMode mode)
         {
-            var url = _urlProviders.Select(provider => provider.GetUrl(_umbracoContext, id, current, mode))
+            string url = _urlProviders.Select(provider => provider.GetUrl(_umbracoContext, id, current, mode))
                 .FirstOrDefault(u => u != null);
             return url ?? "#"; // legacy wants this
         }
 
         internal string GetUrlFromRoute(int id, string route)
         {
-            var provider = _urlProviders.OfType<DefaultUrlProvider>().FirstOrDefault();
-            var url = provider == null 
-                ? route // what else?
-                : provider.GetUrlFromRoute(route, UmbracoContext.Current, id, _umbracoContext.CleanedUmbracoUrl, Mode);
+            IUrlProvider provider = _urlProviders.OfType<IUrlProvider>().FirstOrDefault();
+
+            var url = provider is DefaultUrlProvider
+            ? ((DefaultUrlProvider)provider).GetUrlFromRoute(route, UmbracoContext.Current, id, _umbracoContext.CleanedUmbracoUrl, Mode)
+            : provider.GetUrl(UmbracoContext.Current, id, _umbracoContext.CleanedUmbracoUrl, Mode);

```

Comment:
There's no null checking and fallback to the route value anymore! To keep it compact, a null-conditional and coalescing operator can be combined:
```suggestion
            : provider?.GetUrl(UmbracoContext.Current, id, _umbracoContext.CleanedUmbracoUrl, Mode) ?? route;
```
But for readability, this should probably be restructured...

Reply:
Also, why is the singleton `UmbracoContext.Current` used, if there's also a field `_umbracoContext`?

Reply:
Hi @ronaldbarendse and @pbres 

I think the comments of @ronaldbarendse make sense and should be fixed.

Can you pick that up and after that I will do a test.

Dave

Reply:
Hi @dawoe,

sure thing, thank you for the review @ronaldbarendse 

Pawel

Reply:
> Also, why is the singleton `UmbracoContext.Current` used, if there's also a field `_umbracoContext`?

Hi @ronaldbarendse, I'm also curious about this one, this is how it was done in the original solution but I can't find a good explanation for it. And about `vars`, it's just some weird auto format option in new VS version, already turned it off :)

Thanks one more time for your insights,

Pawel